deepreg/model/loss/image.py

```diff
@@ -82,3 +84,43 @@ def ssd(y_true: tf.Tensor, y_pred: tf.Tensor) -> tf.Tensor:
     :return: shape = (batch,)
     """
     return tf.reduce_mean(tf.square(y_true - y_pred), axis=[1, 2, 3, 4])
+
+
+def global_mutual_information(y_true: tf.Tensor, y_pred: tf.Tensor) -> tf.Tensor:
+    """
+    differentiable global mutual information loss via Parzen windowing method.
+    reference: https://dspace.mit.edu/handle/1721.1/123142
+
+    :y_true: shape = (batch, dim1, dim2, dim3, ch)
+    :y_pred: shape = (batch, dim1, dim2, dim3, ch)
+    :return: shape = (batch,)
+    """
+    bin_centers = tf.linspace(0.0, 1.1, 23)
+    sigma_ratio = 0.5
+    eps = 1.19209e-07
+
+    sigma = tf.reduce_mean(bin_centers[1:] - bin_centers[:-1]) * sigma_ratio
+    preterm = 1 / (2 * tf.math.square(sigma))
+
+    batch, w, h, z, c = y_true.shape
+    y_true = tf.reshape(y_true, [batch, w * h * z, 1])
+    y_pred = tf.reshape(y_pred, [batch, w * h * z, 1])
+    nb_voxels = y_true.shape[1] * 1.0  # number of voxels
+
+    vbc = bin_centers[None, None, ...]
+
+    I_a = tf.math.exp(-preterm * tf.math.square(y_true - vbc))
+    I_a /= tf.reduce_sum(I_a, -1, keepdims=True)
+    pa = tf.reduce_mean(I_a, axis=1, keepdims=True)
+    I_a_permute = tf.transpose(I_a, (0, 2, 1))
+
+    I_b = tf.math.exp(-preterm * tf.math.square(y_pred - vbc))
+    I_b /= tf.reduce_sum(I_b, -1, keepdims=True)
+    pb = tf.reduce_mean(I_b, axis=1, keepdims=True)
+
+    pa = tf.transpose(pa, (0, 2, 1))
+    papb = tf.keras.backend.batch_dot(pa, pb, axes=(2, 1)) + eps
```

Comment:
maybe we can use pa * pb directly? the dot product should be boardcastable

Reply:
agree. avoid using backend functions if possible which caused compatibility issues before...

Reply:
Personally I prefer not to change if it's not necessary. Use a function here will make the purpose clear, and the readers don't need to think about the shape after broadcasting.

Reply:
> 
> 
> agree. avoid using backend functions if possible which caused compatibility issues before...

all right, it have to be changed if compatible issues exist.

Reply:
acc to https://stackoverflow.com/questions/59361689/redundancies-in-tf-keras-backend-and-tensorflow-libraries
although not many votes, i think it's better to not use backend whenever possible ;)
if you want, leave it there and i can change them tmr

Reply:
> 
> 
> acc to https://stackoverflow.com/questions/59361689/redundancies-in-tf-keras-backend-and-tensorflow-libraries
> although not many votes, i think it's better to not use backend whenever possible ;)
> if you want, leave it there and i can change them tmr

I can fix them, will tell you if I need your help, many thanks!